File: Java/src/main/java/QuestionBank/Solution601_700/Solution661.java
package QuestionBank.Solution601_700;

/**
 * @author windows
 */
public class Solution661 {
    public int[][] imageSmoother(int[][] img) {
        int[][] res = new int[img.length][img[0].length];
        for (int i = 0; i < img.length; i++) {
            for (int j = 0; j < img[i].length; j++) {
                int sum = img[i][j];
                int count = 1;
                if (i - 1 >= 0 && j - 1 >= 0){
                    sum += img[i - 1][j - 1] + img[i - 1][j] + img[i][j - 1];
                    count += 3;
                }
                if (i - 1 >= 0 && j + 1 < img[i].length){
                    sum += img[i - 1][j + 1] + img[i][j + 1];
                    count += 2;
                }
                if (i + 1 < img.length && j - 1 >= 0){
                    sum += img[i + 1][j - 1] + img[i + 1][j];
                    count += 2;
                }
                if(i + 1 < img.length && j + 1 < img[i].length){
                    sum += img[i + 1][j + 1];
                    count += 1;
                }
                res[i][j] = sum / count;
            }
        }
        return res;
    }


    public static void main(String[] args) {
        Solution661 solution661 = new Solution661();
    }
}

File: C++/src/Solution601_700/Solution661.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution661.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

vector<vector<int>> Solution661::imageSmoother(vector<vector<int>> &img) {
    int m = img.size();
    int n = img[0].size();
    vector<vector<int>> res(m, vector<int>(n));
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            int num = 0, sum = 0;
            for (int k = i - 1; k <= i + 1; ++k) {
                for (int l = j - 1; l <= j + 1; ++l) {
                    if (k >= 0 && k < m && l >= 0 && l < n){
                        num++;
                        sum += img[k][l];
                    }
                }
            }
            res[i][j] = sum / num;
        }
    }
    return res;
}
